jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,660 @@
+      for (std::size_t i = 0; i != rel_inside + 1; ++i) {
+        // move out and clear the price (if valid)
+        if (top_levels_.at(i) != 0) {
+          price4_t px_i = relative_to_price(i);
+          bottom_levels_.emplace(px_i, top_levels_.at(i));
+          top_levels_.at(i) = 0;
+        }
+      }
+      return;
+    }
+    // px_max is worse than px_inside
+    // prices to move out are from 0.. relative (px_max) excluded
+    auto rel_px_max = price_to_relative(px_max);
+    for (std::size_t i = 0; i != rel_px_max; ++i) {
+      // ... move out the price i (if valid)
+      if (top_levels_.at(i) != 0) {
+        price4_t px_i = relative_to_price(i);
+        bottom_levels_.emplace(px_i, top_levels_.at(i));
+        top_levels_.at(i) = 0;
+      }
+    }
+    // shift price down rel_px_max positions
+    // from px_max to px_inside both included...
+    // ...(the are no better prices than px_inside_)
+    auto rel_px_inside = price_to_relative(px_inside_);
+    std::size_t j = 0;
+    for (std::size_t i = rel_px_max; i <= rel_px_inside; ++i, ++j) {
+      if (top_levels_.at(i) != 0) {
+        top_levels_.at(j) = top_levels_.at(i);
+        top_levels_.at(i) = 0;
+      }
+    }
+  }
+
+  /**
+   * Move relative prices from best price to px_begin_top* (included)
+   * out of bottom_levels_ to top_levels_.
+   * *px_begin_top_ is updated before calling this function.
+   */
+  void move_bottom_to_top() {
+    if (bottom_levels_.empty()) {
+      return; // nothing to move in
+    }
+    // traverse all bottom_levels_ prices worse than or equal to px_begin_top_
+    auto le = bottom_levels_.begin();
+    for (/* */; le != bottom_levels_.end(); ++le) {
+      if (not better_(px_begin_top_, le->first)) {
+        // move price to top_levels_
+        auto rel_px = price_to_relative(le->first);
+        top_levels_.at(rel_px) = le->second;
+      } else {
+        // no more good prices to move...
+        break;
+      }
+    }
+    // erase the moved prices..
+    bottom_levels_.erase(bottom_levels_.begin(), le);
+  }
+
+  /**
+   * @returns the best next price from the inside.
+   * (return empty quote price if none available).
+   */
+  price4_t next_best_price() const {
+    // begins checking below the current inside
+    auto rel_inside = price_to_relative(px_inside_);
+    for (int i = rel_inside - 1; i != -1; --i) {
+      if (top_levels_.at(i) != 0) {
+        // got the next one...
+        return relative_to_price(i);
+      }
+    }
+    return empty_quote().first;
+  }
+
+private:
+  /// function object for inequality comparison of template parameter type.
+  /// if better_(p1,p2) is true means p1 is strictly a better price than p2
+  compare_t better_;
+
+  /// top_levels_ max size
+  std::size_t max_size_;
+
+  /// the best relative prices and quantity
+  std::vector<int> top_levels_;
+
+  /// the worst (tail) absolute prices and quantity
+  std::map<price4_t, int, compare_t> bottom_levels_;
```

Comment:
To me, the obvious way to implement this class would be to have a `std::map<int,int, ...>`, where the index into the map represents the price level (not the price).  So, for example, if the class receives a price of $1.01 that gets immediately converted to the level of 10001, and you only operate on 10001 from that point forward, and only covert back to prices when returning things like the best_quote().

I think that would simplify a lot of your code, where you are converting to prices and back.